Add Create a backend add-on section in the Developer guide with `GenericSetup` (#1757)
plone/documentation: docs/developer-guide/create-a-backend-add-on.md

---
myst:
  html_meta:
    "description": "How to create a backend add-on"
    "property=og:description": "How to create a backend add-on"
    "property=og:title": "Create a backend add-on"
    "keywords": "Plone 6, backend, add-on"
---

(create-a-backend-add-on-label)=

# Create a backend add-on

This section explains how a developer can create an {term}`add-on` for the Plone backend.

## System requirements

Follow the section {ref}`create-project-cookieplone-system-requirements` to set up your system.

## Generate the add-on project with `plonecli`

Choose your desired local development base folder and run the following command to create an addon project with `plonecli`

```shell
uvx plonecli create addon <addon namespace package>
```

```console

> > uvx plonecli create addon collective.addon

Reviewer: I tried this and here's what happened:
```
$ uvx plonecli create addon collective.addon
Traceback (most recent call last):
  File "/Users/davisagli/.local/share/uv/tools/plonecli/bin/plonecli", line 4, in <module>
    from plonecli.cli import cli
  File "/Users/davisagli/.local/share/uv/tools/plonecli/lib/python3.14/site-packages/plonecli/__init__.py", line 4, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'
```

- uv picked the latest Python it knows about, which is 3.14. Does plonecli support it?
- it got the new version of setuptools which doesn't have pkg_resources

It looks like @erral already fixed the latter in https://github.com/plone/plonecli/commit/a00e960e9507bd2a7e1caa835a9bef3d0781fad0 ... could that be released?

Reviewer 2: Yes sorry, I didn't find time to request releases of all of them (mrbob, plonecli, bobtemplates.plone). I'll do it later today.

Reviewer 2: This is already working.

RUN: bobtemplates.plone:addon -O collective.addon

Welcome to mr.bob interactive mode. Before we generate directory structure,
some questions need to be answered.

Answer with a question mark to display help.
Values in square brackets at the end of the questions show the default value if
there is no answer.

> 
> --> Package description [An add-on for Plone]:
> 
> --> Plone version [6.0.0]:
> 
> --> Python version for virtualenv [python3]:
> 
> --> Do you want me to activate VS Code support? (y/n) [y]:
> 
> 
> 
> isort-apply: successful:
> isort-apply: install_deps> python -I -m pip install isort -c constraints.txt
> isort-apply: commands[0]> isort /Users/<username>/Development/collective.addon/src
> /Users/<username>/Development/collective.addon/setup.py
> Fixing /Users/<username>/Development/collective.addon/src/collective/addon/testing.py
> Fixing /Users/<username>/Development/collective.addon/src/collective/addon/tests/test_setup.py
>   isort-apply: OK (2.57=setup[1.94]+cmd[0.63] seconds)
>   congratulations :) (2.59 seconds)
> 
> 
> Identified `/` as project root containing a file system root.
> Sources to be formatted: "Users/<username>/Development/collective.addon/src",
>   "Users/<username>/Development/collective.addon/setup.py"
> src/collective/__init__.py wasn't modified on disk since last run.
> src/collective/addon/browser/__init__.py wasn't modified on disk since last run.
> src/collective/addon/locales/__init__.py wasn't modified on disk since last run.
> src/collective/addon/tests/__init__.py wasn't modified on disk since last run.
> src/collective/addon/interfaces.py already well formatted, good job.
> reformatted src/collective/addon/__init__.py
> reformatted src/collective/addon/setuphandlers.py
> reformatted src/collective/addon/testing.py
> reformatted setup.py
> reformatted src/collective/addon/locales/update.py
> reformatted src/collective/addon/tests/test_setup.py
> 
> All done! ✨ 🍰 ✨
> 6 files reformatted, 5 files left unchanged.
> 
> black-enforce: successful:
> black-enforce: install_deps> python -I -m pip install black -c constraints.txt
> black-enforce: commands[0]> black -v src setup.py
>   black-enforce: OK (2.60=setup[2.12]+cmd[0.48] seconds)
>   congratulations :) (2.61 seconds)
> 
> 
> git init is disabled!
> Generated file structure at /Users/<username>/Development/collective.addon/collective.addon
> ```

Reviewer 2: ```suggestion
RUN: bobtemplates.plone:addon -O collective.addon

Welcome to mr.bob interactive mode. Before we generate directory structure, some questions need to be answered.

Answer with a question mark to display help.
Values in square brackets at the end of the questions show the default value if there is no answer.

--> Author's name [Mikel Larreategi]: 

--> Author's email [mlarreategi@codesyntax.com]: 

--> Author's GitHub username: 

--> Package description [An add-on for Plone]: 

--> Do you want me to initialize a GIT repository in your new package? (y/n) [y]: 

--> Plone version [6.0.0]: 

--> Python version for virtualenv [python3]: 

--> Do you want me to activate VS Code support? (y/n) [y]: 

╭──────────────────────────────── cookieplone ─────────────────────────────────╮
│                                                                              │
│                                   *******                                    │
│                               ***************                                │
│                             ***             ***                              │
│                           ***    ***          ***                            │
│                          ***    *****          ***                           │
│                         ***      ***            ***                          │
│                         ***               ***   ***                          │
│                         ***              *****  ***                          │
│                         ***      ***      ***   ***                          │
│                          ***    *****          ***                           │
│                           ***    ***          ***                            │
│                             ***             ***                              │
│                               ***************                                │
│                                   *******                                    │
│                                                                              │
╰──────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────── Plone Addon ─────────────────────────────────╮
│                                                                              │
│ Creating a new Plone Addon                                                   │
│                                                                              │
│ Sanity check results:                                                        │
│                                                                              │
│   - Cookieplone: ✓                                                           │
│   - uv: ✓                                                                    │
│   - git: ✓                                                                   │
│                                                                              │
╰──────────────────────────────────────────────────────────────────────────────╯
 -> Remove files used in classic UI setup
 -> Create namespace packages
 -> Format code
 -> Initialize Git repository
 -> Generate documentation scaffold
 -> Setup VSCode configuration
 -> Setup GitHub CI
╭────────────────────────── New addon was generated ───────────────────────────╮
│                                                                              │
│ collective.addon                                                             │
│                                                                              │
│ Now, enter the repository, start coding, and push to your organization.      │
│                                                                              │
│ Sorry for the convenience,                                                   │
│ The Plone Community.                                                         │
│                                                                              │
│ https://plone.org/                                                           │
╰──────────────────────────────────────────────────────────────────────────────╯
Done
Generated file structure at /tmp/collective.addon
```

Author: Thanks @erral! The terminal output is very wide, though. The code block area when rendered in the docs starts to wrap at the 90th character, requiring users to horizontal scroll back and forth to view the output. Would you please narrow your terminal session window and run it again?

Reviewer 2: OK, let's try again

╭────────────────────────── New addon was generated ───────────────────────────╮
│                                                                              │
│ Addon                                                                        │
│                                                                              │
│ Now, enter the repository, start coding, and push to your organization.      │
│                                                                              │
│ Sorry for the convenience,                                                   │
│ The Plone Community.                                                         │
│                                                                              │
│ https://plone.org/                                                           │
╰──────────────────────────────────────────────────────────────────────────────╯
```

Cookieplone creates a folder with the name of the add-on, in this example, `collective.addon`.

You can now continue to add subtemplates to your addon {ref}`create-a-backend-add-on-add-subtemplate-label`

(create-a-backend-add-on-add-subtemplate-label)=

## Add `plonecli` subtemplate to an addon

> The generated addon contains a {file}`bobtemplates.cfg` file which lets you add several subtemplates with `plonecli`.

Reviewer 2: ```suggestion
The generated addon contains a {file}`bobtemplate.cfg` file which lets you add several subtemplates with `plonecli`.
```

---
myst:
  html_meta:
    "description": "How to create a backend add-on"
    "property=og:description": "How to create a backend add-on"
    "property=og:title": "Create a backend add-on"
    "keywords": "Plone 6, backend, add-on"
---

(create-a-backend-add-on-label)=

# Create a backend add-on

This section explains how a developer can create an {term}`add-on` for the Plone backend.

> You will use a framework called {term}`GenericSetup` and a tool called {term}`Buildout`.

Reviewer: I don't quite get your vision for this section yet.

Most of the documentation that can be written about how to create a backend add-on will be relevant regardless of how Plone was installed, and also regardless of whether the add-on was created from a Cookieplone template, bobtemplates.plone, or by hand.

So mentioning buildout here is off topic.

And learning about `GenericSetup` is a piece of what goes into creating an add-on, but it's certainly not the only or the most important piece.

Author: I created this file to jump start something for @petschki to work on. I have no vision about this whatsoever. If the snippets aren't needed, they can be tossed out.

I moved the snippets here from another document that had them as headings in a how-to guide, instead of in a conceptual guide. IOW no one does "How to Buildout and GenericSetup".

│                                                                              │
│ Addon                                                                        │
│                                                                              │
│ Now, enter the repository, start coding, and push to your organization.      │
│                                                                              │
│ Sorry for the convenience,                                                   │
│ The Plone Community.                                                         │
│                                                                              │
│ https://plone.org/                                                           │
╰──────────────────────────────────────────────────────────────────────────────╯
```

Cookieplone creates a folder with the name of the add-on, in this example, `collective.addon`.

You can now continue to add subtemplates to your addon {ref}`create-a-backend-add-on-add-subtemplate-label`

## Generate the add-on project with `plonecli`

Run the following command to create an addon project with `plonecli`

```shell
uvx plonecli create addon <addon namespace package>

> ```

Reviewer 2: I would remove using plonecli to create the addon. The latests alpha release calls `uvx cookieplone` under the hood, so it does the same.

We keep the `addon` template in there because all the testing is based on creating the addon

  black-enforce: OK (2.60=setup[2.12]+cmd[0.48] seconds)
  congratulations :) (2.61 seconds)

git init is disabled!
Generated file structure at /Users/<username>/Development/collective.addon/collective.addon
```

Plonecli creates a folder with the name of the add-on, in this example, `collective.addon`.

You can now continue to add subtemplates to your addon {ref}`create-a-backend-add-on-add-subtemplate-label`

(create-a-backend-add-on-add-subtemplate-label)=

## Add `plonecli` subtemplate to an addon

The generated addon contains a {file}`bobtemplates.cfg` file which lets you add several subtemplates with `plonecli`.

Run the following command to list the available subtemplates.

```shell
uvx plonecli -l
```

```shell
> uvx plonecli -l
Available mr.bob templates:
 - addon
  - behavior
  - content_type
  - controlpanel
  - form
  - indexer
  - mockup_pattern
  - portlet
  - restapi_service
  - site_initialization
  - subscriber
  - svelte_app
  - theme
  - theme_barceloneta
  - theme_basic
  - upgrade_step
  - view
  - viewlet
  - vocabulary
 - buildout
 ```

 All templates below `addon` can be added to your newly created addon with:

 ```shell

>  uvx ploncli add <subtemplate>

Reviewer 2: ```suggestion
 uvx plonecli add <subtemplate>
```

 ```

> 

Reviewer 2: You can also use `make add <subtemplate>` inside the `backend` folder of the structure created by cookieplone

Reviewer 3: the Makefile only exists in the monorepo `backend` folder, am I correct? 

If you simply create a `uvx cookieplone backend_addon` there is no Makefile (at least for me) ... I'll outline both scenarios.

Reviewer 2: I have just made a test, I see the correct `Makefile` and it contains the `make add` command support.

Reviewer 3: Oh, sorry... you're right. I've looked in my `plonecli` created addon, not in the `cookieplone backend_addon` version 🫣

I'll update this ...
